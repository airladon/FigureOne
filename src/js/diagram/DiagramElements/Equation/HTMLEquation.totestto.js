// import HTMLEquation from './HTMLEquation';

// describe('HTML Equation', () => {
//   test('Text', () => {
//     const t = new e.Text('test', 'id1', 'c1');
//     const rendered = t.render();
//     const expected = '<div id="id1" class="equation_element c1 equation_text">\n  test\n</div>';
//     expect(rendered).toBe(expected);
//   });
//   test('Line', () => {
//     const t1 = new e.Text('a+b', 'id1', 'c1');
//     const t2 = new e.Text('=', 'id2', 'c2');
//     const t3 = new e.Text('c', 'id3', 'c3');

//     const expected =
// `<div id="id_e1" class="equation_element class_e1">
//   <div id="id1" class="equation_element c1 equation_text">
//     a+b
//   </div>
//   <div id="id2" class="equation_element c2 equation_text">
//     =
//   </div>
//   <div id="id3" class="equation_element c3 equation_text">
//     c
//   </div>
// </div>`;
//     const line = new e.Line(
//       [t1, t2, t3],
//       'id_e1',
//       'class_e1',
//     );
//     expect(line.render()).toBe(expected);
//   });

//   test('Fraction', () => {
//     const n = new e.Line(
//       [new e.Text('a + b')],
//       'id_n1',
//       'class_n1',
//     );
//     const d = new e.Line(
//       [new e.Text('c')],
//       'id_d1',
//       'class_d1',
//     );
//     const f = new e.Fraction(
//       n,
//       d,
//       'id_f1',
//       'class_f1',
//     );
//     const expected =
// `<div id="id_f1" class="equation_element class_f1 fraction">
//   <div class="numerator">
//     <div id="id_n1" class="equation_element class_n1">
//       <div class="equation_element equation_text">
//         a + b
//       </div>
//     </div>
//   </div>
//   <div class="fraction_line"> </div>
//   <div class="denominator">
//     <div id="id_d1" class="equation_element class_d1">
//       <div class="equation_element equation_text">
//         c
//       </div>
//     </div>
//   </div>
// </div>`;
//     expect(f.render()).toBe(expected);
//   });
//   describe('Equation', () => {
//     test('a + b = c', () => {
//       // const line = new e.Line();
//       const eq = new e.Equation('e1', [new e.Text('a + b = c')]);
//       const expected =
// `<div id="e1" class="equation_element equation">
//   <div class="equation_element equation_text">
//     a + b = c
//   </div>
// </div>`;
//       expect(eq.render()).toBe(expected);
//     });
//   });

//   test.only('temp', () => {
//     // console.log(e.e([e.e('b'), e.e('&#177;'), e.sqrt('4ac')]).render());
//     const n = e.e('b');
//     const d = e.e('2');
//     const eq = new e.Equation([e.frac(n, d), e.e('='), e.e('0')]);
//     console.log(eq.render());
//     expect(true).toBe(true);
//   })
// });

// // b +- sqrt(4ac)
// // -------------- = d
// //     2A

// // E = e.e([
// //   e.frac([e.e('b') e.e('+') e.sqrt('4ac')], e.e('2A')),
// //   e.('='),
// //   e.('d')])

// //  / inf
// //  |
// //  | x^2 dx = 2x
// //  |
// // / 0


// // E = e([e.o('&int;').ss('0', 'inf') e.t('x').sp('2') e.t('dx') e.('=') e.('2x')])
