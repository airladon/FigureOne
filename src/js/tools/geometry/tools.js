// @flow
import { Point } from './Point';

function pointsToNumbers2(points: Array<Point>) {
  const list = Array(points.length * 2);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 2] = points[i].x;
    list[i * 2 + 1] = points[i].y;
  }
  return list;
}

function pointsToNumbers(points: Array<Point>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return pointsToNumbers2(points);
  }
  const list = Array(points.length * 3);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 3] = points[i].x;
    list[i * 3 + 1] = points[i].y;
    list[i * 3 + 2] = points[i].z;
  }
  return list;
}

function numbersToPoints2(numbers: Array<number>) {
  const points = Array(numbers.length / 2);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(numbers[i * 2], numbers[i * 2 + 1]);
  }
  return points;
}

function numbersToPoints(numbers: Array<number>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return numbersToPoints2(numbers);
  }
  const points = Array(numbers.length / 3);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(
      numbers[i * 3],
      numbers[i * 3 + 1],
      numbers[i * 3 + 2],
    );
  }
  return points;
}

function toNumbers(
  pointsOrNumbers: Array<Point> | Array<Array<number>> | Array<number>,
  dimension: 2 | 3 = 3,
) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] === 'number') {
    return pointsOrNumbers;
  }
  if (Array.isArray(pointsOrNumbers[0])) {
    const out = [];
    for (let i = 0; i < pointsOrNumbers.length; i += 1) {
      out.push(...pointsOrNumbers);
    }
    return out;
  }
  return pointsToNumbers(pointsOrNumbers, dimension);
}

function toPoints(pointsOrNumbers: Array<Point> | Array<number>, dimension: 2 | 3 = 3) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] !== 'number') {
    return pointsOrNumbers;
  }
  return numbersToPoints(pointsOrNumbers, dimension);
}

export {
  pointsToNumbers2,
  pointsToNumbers,
  numbersToPoints,
  numbersToPoints2,
  toNumbers,
  toPoints,
};
