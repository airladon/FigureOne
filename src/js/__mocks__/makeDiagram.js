// import {
//   DiagramElementPrimitive,
//   DiagramElementCollection,
//   // AnimationPhase,
// } from '../Element';
import Diagram from '../diagram/Diagram';
import {
  Rect, getPoint, Point,
} from '../tools/g2';
import webgl from './WebGLInstanceMock';
import DrawContext2D from './DrawContext2DMock';
// import VertexPolygon from '../DrawingObjects/VertexObject/VertexPolygon';
import * as tools from '../tools/tools';
import { round } from '../tools/math';

tools.isTouchDevice = jest.fn();

jest.mock('../diagram/Gesture');
jest.mock('../diagram/webgl/webgl');
jest.mock('../diagram/DrawContext2D');
jest.mock('../diagram/DrawingObjects/HTMLObject/HTMLObject');

const generateRandomStringMock = jest.spyOn(tools, 'generateRandomString');
generateRandomStringMock.mockImplementation(() => '000000');

export default function makeDiagram(
  inputCanvas = new Rect(100, -300, 1000, 500),
  inputLimits = new Rect(-1, -1, 2, 2),
) {
  document.body.innerHTML =
    '<div id="c">'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__low">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__low">'
    + '  </canvas>'
    + '  <div class="figureone__html">'
    + '  </div>'
    + '  <canvas class="figureone__gl" id="id_figureone__gl__high">'
    + '  </canvas>'
    + '  <canvas class="figureone__text" id="id_figureone__text__high">'
    + '  </canvas>'
    + '</div>';
  // canvas = document.getElementById('c');
  const definition = {
    width: inputCanvas.width,
    height: inputCanvas.height,
    limits: inputLimits,
  };

  const canvasMock = {
    width: definition.width,
    height: definition.height,
    // offsetLeft: 100,
    left: inputCanvas.left,
    // offsetTop: 200,
    top: inputCanvas.top,
    // width: definition.width,
    // height: definition.height,
    offsetWidth: definition.width,
    offsetHeight: definition.height,
    scrollLeft: 0,
    scrollTop: 0,
    // eslint-disable-next-line arrow-body-style
    getBoundingClientRect: () => {
      return {
        left: inputCanvas.left,
        top: inputCanvas.top,
        width: definition.width,
        height: definition.height,
      };
    },
  };
  const htmlCanvasMock = {
    style: {
      fontsize: 1,
    },
    offsetWidth: 100,
    appendChild: () => {},
  };
  const { limits } = definition;
  const diagram = new Diagram({ htmlId: 'c', limits });
  diagram.webglLow = webgl;
  diagram.webglHigh = webgl;
  diagram.webgl = webgl;
  diagram.canvasLow = canvasMock;
  diagram.canvasHigh = canvasMock;
  diagram.htmlCanvas = htmlCanvasMock;
  diagram.isTouchDevice = false;
  diagram.draw2DLow = new DrawContext2D(definition.width, definition.height);
  diagram.draw2DHigh = new DrawContext2D(definition.width, definition.height);
  diagram.draw2D = diagram.draw2DLow;
  diagram.shapesLow = diagram.getShapes(false);
  diagram.shapesHigh = diagram.getShapes(true);
  diagram.shapes = diagram.shapesLow;
  diagram.equationLow = diagram.getEquations(false);
  diagram.equationHigh = diagram.getEquations(true);
  diagram.equation = diagram.equationLow;
  diagram.advancedLow = diagram.getObjects(false);
  diagram.advancedHigh = diagram.getObjects(true);
  diagram.advanced = diagram.advancedLow;
  diagram.setSpaceTransforms();
  // needed as the first element needs to be set with the space Transforms
  diagram.initElements();
  diagram.mock = {
    initialTime: 0,
    duration: 0,
    previousTouchPoint: new Point(0, 0),
    timersBeforeDraw: true,
    timeStep: (deltaTimeInSecondsIn, frameTimeIn = null) => {
      const { duration, initialTime } = diagram.mock;
      const deltaTimeInSeconds = round(deltaTimeInSecondsIn, 8);
      let frameTime = deltaTimeInSeconds;
      // let deltaTime = 0;
      if (frameTimeIn != null && frameTimeIn < deltaTimeInSeconds) {
        frameTime = frameTimeIn;
        // deltaTime = frameTime;
      }
      let deltaTime = frameTime;

      // let delta = Math.min(deltaTimeInSeconds, frameTime);
      let lastTime = 0;
      deltaTime = round(deltaTime, 8);
      while (deltaTime <= round(deltaTimeInSeconds, 8)) {
        // if (window.asdf) {
        //   console.log(duration, deltaTime + 0.00);
        // }
        const newNow = round((duration + deltaTime + initialTime) * 1000, 8);
        global.performance.now = () => newNow;
        if (diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        diagram.animateNextFrame();
        diagram.draw(round(newNow / 1000, 8));
        if (!diagram.mock.timersBeforeDraw) {
          jest.advanceTimersByTime(round((deltaTime - lastTime) * 1000, 8));
        }
        lastTime = deltaTime;
        if (deltaTime === deltaTimeInSeconds || frameTime === 0) {
          deltaTime += 1;
        } else if (deltaTime + frameTime > deltaTimeInSeconds) {
          deltaTime = deltaTimeInSeconds;
        } else {
          deltaTime += frameTime;
        }
        deltaTime = round(deltaTime, 8);
      }
      diagram.mock.duration += deltaTimeInSeconds;
    },
    touchDown: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchDownHandler(clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
    touchUp: () => {
      diagram.touchUpHandler();
    },
    touchMove: (diagramPosition) => {
      const p = getPoint(diagramPosition);
      const pixelPoint = p.transformBy(diagram.spaceTransforms.diagramToPixel.m());
      const clientPoint = diagram.pixelToClient(pixelPoint);
      diagram.touchMoveHandler(diagram.mock.previousTouchPoint, clientPoint);
      diagram.mock.previousTouchPoint = clientPoint;
    },
  };
  diagram.globalAnimation.reset();
  return diagram;
}

