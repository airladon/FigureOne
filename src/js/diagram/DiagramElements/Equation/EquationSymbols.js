// @flow
import {
  Point, Transform,
} from '../../../tools/g2';
import { joinObjects } from '../../../tools/tools';
import DiagramPrimatives from '../../DiagramPrimatives/DiagramPrimatives';
import Integral from './Symbols/Integral';
// import SuperSub from './Elements/SuperSub';
import Bracket from './Symbols/Bracket';
import RoundedSquareBracket from './Symbols/RoundedSquareBracket';
import Bar from './Symbols/Bar';
import Brace from './Symbols/Brace';
import SquareBracket from './Symbols/SquareBracket';
// import { Annotation, AnnotationInformation } from './Elements/Annotation';

export default class EquationSymbols {
  shapes: DiagramPrimatives;
  defaultColor: Array<number>;

  constructor(
    shapes: DiagramPrimatives,
    defaultColor: Array<number>,
  ) {
    this.shapes = shapes;
    this.defaultColor = defaultColor;
  }

  get(
    name: string,
    options: {
      color?: Array<number>,
      numLines?: number,
      side?: 'left' | 'right' | 'bottom' | 'top',
    },
  ) {
    if (name === 'vinculum') {
      return this.vinculum(options);
    }
    if (name === 'strike') {
      return this.strike(options);
    }
    if (name === 'xStrike') {
      return this.xStrike(options);
    }
    if (name === 'integral') {
      return this.integral(options);
    }
    if (name === 'bracket') {
      return this.bracket(options);
    }
    if (name === 'squareBracket') {
      return this.squareBracket(options);
    }
    if (name === 'brace') {
      return this.brace(options);
    }
    if (name === 'bar') {
      return this.bar(options);
    }
    if (name === 'roundedSquareBracket') {
      return this.roundedSquareBracket(options);
    }
    return null;
  }

  vinculum(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('vinculum').scale(1, 1).translate(0, 0),
    );
  }

  strike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    return this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strike').scale(1, 1).rotate(0).translate(0, 0),
    );
  }

  xStrike(options: { color?: Array<number> } = {}) {
    let { color } = options;
    if (color == null) {
      color = this.defaultColor;
    }
    const cross = this.shapes.collection(new Transform('xStrike').scale(1, 1).rotate(0).translate(0, 0));
    cross.color = color;
    const strike1 = this.shapes.horizontalLine(
      new Point(0, 0),
      1, 1, 0,
      color,
      new Transform('strikeLine').scale(1, 1).rotate(0).translate(0, 0),
    );
    const strike2 = strike1._dup();
    cross.add('s1', strike1);
    cross.add('s2', strike2);
    return cross;
  }

  integral(options: { color?: Array<number>, numLines?: number}) {
    const defaultOptions = {
      color: this.defaultColor,
      numLines: 1,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Integral(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.numLines,
      new Transform('integral').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }


  bracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    numLines?: number,
    color?: Array<number>,
  }) {
    const defaultOptions = {
      side: 'left',
      numLines: 1,
      color: this.defaultColor,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Bracket(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.side,
      optionsToUse.numLines,
      new Transform('bracket').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }


  bar(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    numLines?: number,
    color?: Array<number>,
  }) {
    const defaultOptions = {
      side: 'top',
      numLines: 1,
      color: this.defaultColor,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Bar(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.side,
      optionsToUse.numLines,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }

  squareBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    numLines?: number,
    color?: Array<number>,
  }) {
    const defaultOptions = {
      side: 'left',
      numLines: 1,
      color: this.defaultColor,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new SquareBracket(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.side,
      optionsToUse.numLines,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }

  brace(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    numLines?: number,
    color?: Array<number>,
  }) {
    const defaultOptions = {
      side: 'left',
      numLines: 1,
      color: this.defaultColor,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new Brace(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.side,
      optionsToUse.numLines,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }

  roundedSquareBracket(options: {
    side?: 'left' | 'right' | 'top' | 'bottom',
    numLines?: number,
    color?: Array<number>,
  }) {
    const defaultOptions = {
      side: 'left',
      numLines: 1,
      color: this.defaultColor,
    };
    const optionsToUse = joinObjects(defaultOptions, options);
    return new RoundedSquareBracket(
      this.shapes.webgl,
      optionsToUse.color,
      optionsToUse.side,
      optionsToUse.numLines,
      new Transform('bar').scale(1, 1).translate(0, 0),
      this.shapes.limits,
    );
  }
}
