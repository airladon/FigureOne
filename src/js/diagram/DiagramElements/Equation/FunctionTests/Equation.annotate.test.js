// import {
//   Point,
// } from '../../../../tools/g2';
import {
  round,
} from '../../../../tools/math';
import * as tools from '../../../../tools/tools';
import makeDiagram from '../../../../__mocks__/makeDiagram';
import { EquationNew } from '../Equation';

tools.isTouchDevice = jest.fn();

jest.mock('../../../Gesture');
jest.mock('../../../webgl/webgl');
jest.mock('../../../DrawContext2D');

describe('Equation Functions - Annotations', () => {
  let diagram;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    diagram = makeDiagram();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      c: 'c',
      d: 'd',
      e: 'e',
      f: 'f',
      g: 'g',
    };
    functions = {
      single: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const annotate = e.annotate.bind(e);
        eqn.addElements(elements);
        const annotation = ['b', 'right', 'bottom', 'left', 'top', 0.5];
        const annotations = [annotation];

        eqn.addForms({
          // Full Object
          0: {
            content: {
              annotate: {
                content: 'a',
                withAnnotations: annotations,
              },
            },
          },
          //   // Method Object
          1: {
            annotate: {
              content: 'a',
              withAnnotations: annotations,
            },
          },
          // Method Array
          2: {
            annotate: ['a', annotations],
          },
          // Function with Method Array
          3: e.annotate(['a', annotations]),
          // Function with parameters
          4: e.annotate('a', annotations),
          // Bound Function with parameters
          5: annotate('a', annotations),
          // Bound Function with Object
          6: annotate({
            content: 'a',
            withAnnotations: annotations,
          }),
          7: {
            annotate: {
              content: 'a',
              withAnnotations: {
                annotation: {
                  annotation: 'b',
                  relativeToContent: ['right', 'bottom'],
                  relativeToAnnotation: ['left', 'top'],
                  scale: 0.5,
                },
              },
            },
          },
          8: {
            annotate: {
              content: 'a',
              withAnnotations: ['b', 'right', 'bottom', 'left', 'top', 0.5],
            },
          },
          // 7: {
          //   annotate: ['a', annotation],
          // },
        });
      },
      annotations: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const annotation = e.annotation.bind(e);
        const annotate = e.annotate.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Method Object
          0: annotate('a', [
            {
              annotation: {
                annotation: 'b',
                relativeToContent: ['right', 'bottom'],
                relativeToAnnotation: ['left', 'top'],
                scale: 0.5,
              },
            },
          ]),
          // Method Array
          1: annotate('a', [
            {
              annotation: ['b', 'right', 'bottom', 'left', 'top', 0.5],
            },
          ]),
          // Array only
          2: annotate('a', [
            'b', 'right', 'bottom', 'left', 'top', 0.5,
          ]),
          // Array of array annotations
          3: annotate('a', [[
            'b', 'right', 'bottom', 'left', 'top', 0.5,
          ]]),
          // Function with method Array
          4: annotate('a', [
            e.annotation(['b', 'right', 'bottom', 'left', 'top', 0.5]),
          ]),
          // Function with parameters
          5: annotate('a', [
            e.annotation('b', 'right', 'bottom', 'left', 'top', 0.5),
          ]),
          // Bound function with parameters
          6: annotate('a', [
            annotation('b', 'right', 'bottom', 'left', 'top', 0.5),
          ]),
          // Bound Function with Object
          7: annotate('a', [
            annotation({
              annotation: 'b',
              relativeToContent: ['right', 'bottom'],
              relativeToAnnotation: ['left', 'top'],
              scale: 0.5,
            }),
          ]),
        });
      },
      multiple: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const annotation = e.annotation.bind(e);
        const annotate = e.annotate.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Method Object
          0: annotate('a', [
            {
              annotation: {
                annotation: 'b',
                relativeToContent: ['right', 'bottom'],
                relativeToAnnotation: ['left', 'top'],
                scale: 0.5,
              },
            },
            {
              annotation: {
                annotation: 'c',
                relativeToContent: ['left', 'top'],
                relativeToAnnotation: ['right', 'bottom'],
                scale: 0.5,
              },
            },
          ]),
          // Functions
          1: annotate('a', [
            annotation('b', 'right', 'bottom', 'left', 'top', 0.5),
            annotation('c', 'left', 'top', 'right', 'bottom', 0.5),
          ]),
          // Arrays only
          2: annotate('a', [
            ['b', 'right', 'bottom', 'left', 'top', 0.5],
            ['c', 'left', 'top', 'right', 'bottom', 0.5],
          ]),
        });
      },
      nested: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const annotation = e.annotation.bind(e);
        const annotate = e.annotate.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Method Object
          0: annotate('a', [
            {
              annotation: {
                annotation: annotate(
                  'b', ['c', 'center', 'bottom', 'center', 'top'],
                ),
                relativeToContent: ['right', 'bottom'],
                relativeToAnnotation: ['left', 'top'],
                scale: 0.5,
              },
            },
            {
              annotation: {
                annotation: 'd',
                relativeToContent: ['left', 'top'],
                relativeToAnnotation: ['right', 'bottom'],
                scale: 0.5,
              },
            },
          ]),
          // Functions
          1: annotate('a', [
            annotation(annotate(
              'b', ['c', 'center', 'bottom', 'center', 'top'],
            ), 'right', 'bottom', 'left', 'top', 0.5),
            annotation('d', 'left', 'top', 'right', 'bottom', 0.5),
          ]),
          // Arrays only
          2: annotate('a', [
            [
              annotate(
                'b', ['c', 'center', 'bottom', 'center', 'top'],
              ), 'right', 'bottom', 'left', 'top', 0.5,
            ],
            ['d', 'left', 'top', 'right', 'bottom', 0.5],
          ]),
        });
      },
      parameters: () => {
        eqn = new EquationNew(diagram.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const annotate = e.annotate.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // without
          //   // Method Object
          without: [
            annotate(
              {
                content: 'a',
                withAnnotations: [
                  {
                    annotation: {
                      annotation: 'b',
                      relativeToContent: ['right', 'bottom'],
                      relativeToAnnotation: ['left', 'top'],
                      scale: 0.5,
                    },
                  },
                ],
              },
            ),
            'c',
          ],
          // With parameters
          0: [
            annotate(
              {
                content: 'a',
                withAnnotations: [
                  {
                    annotation: {
                      annotation: 'b',
                      relativeToContent: ['right', 'bottom'],
                      relativeToAnnotation: ['left', 'top'],
                      scale: 0.5,
                    },
                  },
                ],
                includeAnnotationInSize: false,
              },
            ),
            'c',
          ],
          // Method Array
          1: [
            annotate(
              'a',
              ['b', 'right', 'bottom', 'left', 'top', 0.5],
              false,
            ),
            'c',
          ],
          // 1: ['a', { brac: ['b', 'lb', 'rb', 0.1, 0.2, false, 2] }, 'c'],
          // // Function with parameters
          // 2: ['a', brac('b', 'lb', 'rb', 0.1, 0.2, false, 2), 'c'],
        });
      },
    };
  });
  test('Single', () => {
    functions.single();
    const elems = [eqn._a, eqn._b, eqn._c];
    const formsToTest = ['1', '2', '3', '4', '5', '6', '7', '8'];

    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });
  });
  test('Annotations', () => {
    functions.annotations();
    const elems = [eqn._a, eqn._b, eqn._c];
    const formsToTest = ['1', '2', '3', '4', '5', '6', '7'];

    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });
  });
  test('Multiple Annotations', () => {
    functions.multiple();
    const elems = [eqn._a, eqn._b, eqn._c];
    const formsToTest = ['1', '2'];

    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });
  });
  test('Nested Annotations', () => {
    functions.nested();
    const elems = [eqn._a, eqn._b, eqn._c];
    const formsToTest = ['1', '2'];

    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Snapshot test on most complex layout
    eqn.showForm('0');
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._b.transform.mat)).toMatchSnapshot();
    expect(round(eqn._c.transform.mat)).toMatchSnapshot();
    expect(round(eqn._d.transform.mat)).toMatchSnapshot();
  });
  test('Annotation Parameters', () => {
    functions.parameters();
    const elems = [eqn._a, eqn._b, eqn._c];
    const withFormsToTest = ['1'];

    // get without positions
    eqn.showForm('without');
    const withoutPos = elems.map(elem => round(elem.transform.mat).slice());

    // with reference positions
    eqn.showForm('0');
    const withPos = elems.map(elem => round(elem.transform.mat).slice());

    expect(withoutPos).not.toEqual(withPos);

    withFormsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(withPos).toEqual(positions);
    });
  });
});
