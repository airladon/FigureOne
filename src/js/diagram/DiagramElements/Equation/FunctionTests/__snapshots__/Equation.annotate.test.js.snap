// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Annotations Nested Annotations 1`] = `
Array [
  0.7,
  0,
  0.014,
  0,
  0.7,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Annotations Nested Annotations 2`] = `
Array [
  0.35,
  0,
  0.042,
  0,
  0.35,
  -0.02184,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Annotations Nested Annotations 3`] = `
Array [
  0.21,
  0,
  0.0448,
  0,
  0.21,
  -0.03094,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Annotations Nested Annotations 4`] = `
Array [
  0.35,
  0,
  0,
  0,
  0.35,
  0.02772,
  0,
  0,
  1,
]
`;
